feat(delegation): auto associate when possible

MaxMustermann2 opened the pull request:
Whenever the asset bearing `ExocoreAssetID` is delegated by a staker to an operator, we should check if the staker address is the same as the operator address. If a match is found, they should be automatically associated.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Added automatic operator-staker association for self-delegation scenarios
  - Enhanced delegation process with improved association mechanism

- **Testing**
  - Added new test case to validate automatic operator association during self-delegation
  - Updated test suite with refined asset and constant references

- **Improvements**
  - Updated module dependencies to utilize Exocore network types
  - Added clarifying comments for asset information handling related to slashing and undelegation
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
feat(delegation): auto associate when possible

Whenever the asset bearing `ExocoreAssetID` is delegated by a staker to
an operator, we should check if the staker address is the same as the
operator address. If a match is found, they should be automatically
associated.

TimmyExogenous commented on x/delegation/keeper/delegation.go:
```diff
@@ -62,6 +63,13 @@ func (k *Keeper) delegateTo(
 		}
+		// auto associate it, if there is a match. note that both are byte versions of bech32 AccAddress.
+		if bytes.Equal(params.StakerAddress, params.OperatorAddress[:]) {
+			err := k.SetAssociatedOperator(ctx, stakerID, params.OperatorAddress.String())
```
Can we check if it has already been associated to avoid repetitive association?

MaxMustermann2 replied:
Updated the comment as discussed; please check it out.

adu-web3 commented on x/delegation/keeper/delegation.go:
```diff
@@ -62,6 +63,13 @@ func (k *Keeper) delegateTo(
 			return err
 		}
+		// auto associate it, if there is a match. note that both are byte versions of bech32 AccAddress.
+		if bytes.Equal(params.StakerAddress, params.OperatorAddress[:]) {
```
do we still need the associate/dissociate function in delegation precompile in this case? since the evm address that initiates the client chain delegation request could share the same bytes representation with the operator address, and we can require that operator should initiate delegations request from client chain with an evm address sharing the same bytes representation with its exocore bech32 operator address(meaning the same private key)

MaxMustermann2 replied:
I thought the intention for the precompile's association function is to support it on non EVM chains

MaxMustermann2 committed:
doc(delegation): update comment about duplicate

- It is not possible to overwrite an association of the `stakerID`
  ending with `0x0` because such an association cannot be made using any
  other mechanism.
- Even if it exists, the overwriting will do nothing of import due to
  the equality check.

MaxMustermann2 committed:
test(delegation): check operator share auto assoc

Validate that the operator share is correctly set to the expected value
when self-staking ExocoreAssetID

MaxMustermann2 committed:
doc(assets): clarify StakerAssetInfo

- for ExocoreAssetID, the `TotalDepositAmount` is post-slashing, if any
  was applied.
- for other asset IDs, all components are pre-slashing.
